Docker.DotNet.X509/PrivateKey.cs

```diff
@@ -0,0 +1,158 @@
+                }
+
+                endIndex = (int)(mem.Position - endTag.Length - 2);
+            }
+
+            // Convert the bytes from base64;
+            var convertedBytes = Convert.FromBase64String(Encoding.UTF8.GetString(allBytes, startIndex, endIndex - startIndex));
+            mem = new MemoryStream(convertedBytes);
+            using (var rdr = new BinaryReader(mem))
+            {
+                var val = rdr.ReadUInt16();
+                if (val != 0x8230)
+                {
+                    throw new Exception("Invalid byte ordering.");
+                }
+
+                // Discard the next bits of the version.
+                rdr.ReadUInt32();
+                if (rdr.ReadByte() != Padding)
+                {
+                    throw new InvalidDataException("Invalid ASN.1 format.");
+                }
+
+                var rsa = new RSAParameters();
+                rsa.Modulus = rdr.ReadBytes(ReadIntegerCount(rdr));
+                rsa.Exponent = rdr.ReadBytes(ReadIntegerCount(rdr));
+                rsa.D = rdr.ReadBytes(ReadIntegerCount(rdr));
+                rsa.P = rdr.ReadBytes(ReadIntegerCount(rdr));
+                rsa.Q = rdr.ReadBytes(ReadIntegerCount(rdr));
+                rsa.DP = rdr.ReadBytes(ReadIntegerCount(rdr));
+                rsa.DQ = rdr.ReadBytes(ReadIntegerCount(rdr));
+                rsa.InverseQ = rdr.ReadBytes(ReadIntegerCount(rdr));
+
+                // Use "1" to indicate RSA.
+                var csp = new CspParameters(1);
+                
+                // Set the KeyContainerName so that native code that looks up the private key
+                // can find it. This produces a keyset file on disk as a side effect.
+                csp.KeyContainerName = pemFilePath;
+                
+                var rsaProvider = new RSACryptoServiceProvider(csp);
+
+                // Setting to false makes sure the keystore file will be cleaned up
+                // when the current process exits.
+                rsaProvider.PersistKeyInCsp = false;
+                
+                // Import the private key into the keyset.
+                rsaProvider.ImportParameters(rsa);
+
+                return rsaProvider;
+            }
+        }
+
+        /// <summary>
+        /// Reads an integer count encoding in DER ASN.1 format.
+        /// <summary>
+        private static int ReadIntegerCount(BinaryReader rdr)
+        {
+            const byte highBitOctet = 0x80;
+            const byte ASN1_INTEGER = 0x02;
+
+            if (rdr.ReadByte() != ASN1_INTEGER)
+            {
+                throw new Exception("Integer tag expected.");
+            }
+
+            int count = 0;
+            var val = rdr.ReadByte();
+            if ((val & highBitOctet) == highBitOctet)
+            {
+                byte numOfOctets = (byte)(val - highBitOctet);
+                Debug.Assert(numOfOctets <= 4);
```

Comment:
This should be a throw I guess. Its really `untrusted` data and would overflow if this wasn't true.